app/Libraries/Session/Store.php

```diff
@@ -235,63 +159,58 @@ public function getIdWithoutKeyPrefix()
      * @param  string  $sessionId
      * @return bool
      */
-    public function migrate($destroy = false, int $userId = null)
+    public function migrate($destroy = false)
     {
-        // Overriden to allow passing through $userId to namespace session ids
-
         if ($destroy) {
-            if (!$this->isGuestSession()) {
-                static::removeFullId($userId, $this->getId());
+            $userId = $this->userId();
+            if ($userId !== null) {
+                // Keep existing attributes
+                $attributes = $this->attributes;
+                static::batchDelete($userId, [$this->getId()]);
+                $this->attributes = $attributes;
             }
-            $this->handler->destroy($this->getId());
         }
-        $this->setExists(false);
-        $this->setId($this->generateSessionId($userId));
 
-        return true;
+        return parent::migrate($destroy);
     }
 
     /**
      * Save the session data to storage.
+     *
+     * Overriden to track user sessions in Redis and shorten lifetime for guest sessions.
      */
     public function save()
     {
-        $isGuest = $this->isGuestSession();
+        $userId = $this->userId();
 
         if ($this->handler instanceof CacheBasedSessionHandler) {
-            $this->handler->setMinutes($isGuest ? 120 : config('session.lifetime'));
+            $this->handler->setMinutes($userId === null ? 120 : config('session.lifetime'));
         }
 
-        // Overriden to track user sessions in Redis
         parent::save();
 
-        if (!$isGuest) {
-            Redis::sadd(config('cache.prefix').':'.$this->getCurrentKeyPrefix(), $this->getKey());
+        // TODO: move this to migrate and validate session id in readFromHandler
+        if ($userId !== null) {
+            static::redis()->sadd(static::listKey($userId), $this->getId());
         }
     }
 
-    /**
-     * Get a new, random session ID.
-     *
-     * @return string
-     */
-    protected function generateSessionId(int $userId = null)
+    public function userId(): ?int
     {
-        // Overriden to allow namespacing the session id (used as the redis key)
-
-        return static::keyPrefix($userId).':'.Str::random(static::SESSION_ID_LENGTH);
+        // From `Auth::getName()`.
+        // Hardcoded because Auth depends on this class instance which then
+        // calls this functions and would otherwise cause circular dependency.
+        return $this->attributes['login_web_59ba36addc2b2f9401580f014c7f58ea4e30989d'] ?? null;
```

Comment:
should also mention its using the sha1 of the class in `Illuminate\Auth\SessionGuard::getName()` which is always the same 
...I'm pretty sure there'll be other trollings if the hash ever gets changed, anyway 🤔

Reply:
The detail shouldn't matter and I've added relevant test.